test/hotspot/jtreg/compiler/locks/TestNestedSynchronize.java

```diff
@@ -23,216 +23,107 @@
 
 /**
  * @test
- * @bug 8322996
- * @summary Ensure no assert error in C2 with deeply nested synchronize
- *          statements.
- * @run main/othervm -XX:CompileCommand=compileonly,compiler.locks.TestNestedSynchronize::test
+ * @bug 8322996 8324839 8325467
+ * @summary Ensure C2 can compile deeply nested synchronize statements.
+ *          Exercises C2 register masks, in particular. We incrementally
+ *          increase the level of nesting (up to 100) to trigger potential edge
+ *          cases.
+ * @modules java.base/jdk.internal.misc
+ * @library /test/lib /
+ * @run main/othervm -XX:CompileCommand=compileonly,Test::test*
  *                   -Xcomp
+ *                   -XX:+UnlockDiagnosticVMOptions
+ *                   -XX:+AbortVMOnCompilationFailure
  *                   compiler.locks.TestNestedSynchronize
  */
 
 package compiler.locks;
 
+import compiler.lib.compile_framework.*;
+import java.util.ArrayList;
+
 public class TestNestedSynchronize {
+    static int min = 1;
+    static int max = 100;
+    static String test_class_name = "Test";
 
-    public static void main(String[] args) {
-        test();
+    // The below method generates a program of the form:
+    //
+    // public class Test {
+    //     public static void test1() {
+    //         synchronized (Test.class) {
+    //
+    //         }
+    //     }
+    //
+    //
+    //     public static void test2() {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //
+    //         }
+    //         }
+    //     }
+    //
+    //
+    //     public static void test3() {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //
+    //         }
+    //         }
+    //         }
+    //     }
+    //
+    //     ...
+    //
+    //     public static void test99() {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //         ...
+    //
+    //         ...
+    //         }
+    //         }
+    //         }
+    //     }
+    // }
+    //
+    // The above is a massive program. Therefore, we do not directly inline the
+    // program in TestNestedSynchronize and instead compile and run it via the
+    // CompileFramework.
+    public static String generate_test() {
+        ArrayList<String> methods = new ArrayList<String>();
+        for (int i = min; i < max; i++) {
+            String inner = "";
+            for (int j = 0; j < i; j++) {
+                inner = String.format("""
+                          synchronized (%s.class) {
+                  %s
+                          }""", test_class_name, inner);
+            }
```

Comment:
Probably this is not so bad with only 100 repetitions ... but this leads to a lot of successive strings generated, resulting in larger and larger strings allocated, essentially requiring quadratic memory. A StringBuilder would help here. Maybe this is just an FYI: the TemplateFramework internally uses a StringBuilder to avoid this successive concatenation / allocation issue.

Reply:
Thanks for the comment; I implemented it in this way for readability.I refactored it now to avoid the quadratic behavior (assuming `addAll`, `addFirst`, and `addLast` are O(1) for `LinkedList`, and that `String.join` is linear). Please have a look! `StringBuilder` is efficient for appending, but not prepending, right?